src/DI/ContainerBuilder.php

```diff
@@ -463,9 +468,14 @@ private function resolveEntityClass($entity, $recursive = [])
 				throw new ServiceCreationException(sprintf("Factory '%s' used in service '%s' is not callable.", Nette\Utils\Callback::toString($entity), $name[0]));
 			}
 
-			$class = preg_replace('#[|\s].*#', '', PhpReflection::parseAnnotation($reflection, 'return'));
-			if ($class) {
-				$class = $refClass ? PhpReflection::expandClassName($class, $refClass) : ltrim($class, '\\');
+			if (PHP_VERSION_ID >= 70000 && $reflection->hasReturnType()) {
+				$type = $reflection->getReturnType();
+				$class = !$type->isBuiltIn() ? (string) $type : null;
```

Comment:
I hope I didn't misunderstood the code (and it's intent and use cases), but it seems it's just for resolving classes, not every type.

If you add support for resolving builtin type and then have a factory with `createString(): string`, when using it (ie. `@x::createString`), resolving fails with `Nette\DI\ServiceCreationException: Class or interface string used in service 'y' not found.` - and since scalar type can't be a service (or can it?), I find resolving it pointless. Maybe an exception that you can't resolve builtin types should be thrown, though?